mono/mcs/class/System.Windows.Forms/System.Windows.Forms.CocoaInternal/XplatUICocoa.cs

```diff
@@ -1516,7 +1516,14 @@ internal override void PaintEventEnd (ref Message msg, IntPtr handle, bool clien
 		internal override bool PeekMessage(Object queue_id, ref MSG msg, IntPtr hWnd, int wFilterMin, int wFilterMax, uint flags)
 		{
 			bool peeking = 0 == ((uint) PeekMessageFlags.PM_REMOVE & flags);
-			PumpNativeEvent (false, ref msg, false);
+
+			if (!peeking && hWnd == IntPtr.Zero && wFilterMin == 0 && wFilterMax == int.MaxValue)
+			{
+				// NSApp.NextEvent(dequeue:=true) triggers ApplicationShouldTerminate, false does not not
+				return PumpNativeEvent(false, ref msg, true);

```

Comment:
I'm not exactly sure how this affects the overall Cocoa behaviour since `PumpNativeEvent` with `dequeue: true` would normally be called anyway after the filtering condition (at line 1522 in the original code, or 1529 in your code). If the event is not returned by `PumpNativeEvent` or the filtering condition discards it then it would be nice to document that in the comment.

Nevertheless it seems like the change should not break anything and could be a good optimisation, so I agree it makes sense to do it.